histomicstk/segmentation/label/isbfcpp.cpp

```diff
@@ -0,0 +1,255 @@
+/*
+C++ version of ISBF for TraceBounds
+*/
+#include "isbfcpp.h"
+
+#include <iostream>
+#include <list>
+#include <cmath>
+#include <vector>
+
+isbfcpp::isbfcpp(){}
+
+std::vector <std::vector<int> > isbfcpp::rotateMatrix(int nrows, int ncols, std::vector <std::vector<int> > input)
+{
+    std::vector <std::vector<int> > output(ncols, std::vector<int>(nrows));
+    for (int i=0; i<nrows; i++){
+      for (int j=0;j<ncols; j++){
+        output[j][nrows-1-i] = input[i][j];
+      }
+    }
+    return output;
+}
+
+std::vector <std::vector<int> > isbfcpp::traceBoundary(int nrows, int ncols, std::vector <std::vector<int> > mask, int startX, int startY, float inf)
+{
+    // initialize boundary vector
+    std::list<int> boundary_listX;
+    std::list<int> boundary_listY;
+
+    // push the first x and y points
+    boundary_listX.push_back(startX);
+    boundary_listY.push_back(startY);
+
+    // initialize matrix for 0, 90, 180, 270 degrees
+    std::vector <std::vector<int> > matrix00(nrows, std::vector<int>(ncols));
+    std::vector <std::vector<int> > matrix90(ncols, std::vector<int>(nrows));
+    std::vector <std::vector<int> > matrix180(nrows, std::vector<int>(ncols));
+    std::vector <std::vector<int> > matrix270(ncols, std::vector<int>(nrows));
+
+    // copy mask to matrix00
+    for(int i=0; i< nrows; i++){
+        for(int j=0; j< ncols; j++){
+            //matrix00[i][j] = mask[i*ncols+j];
+            matrix00[i][j] = mask[i][j];
+        }
+    }
+
+    // rotate matrix for 90, 180, 270 degrees
+    matrix270 = rotateMatrix(nrows, ncols, matrix00);
```

Comment:
returning a vector from rotateMatrix and assigning it will result in a copy. You may want to pass in a reference to the output matrix as a function argument to speed up.